### PR TITLE
[Storage] Client side encryption with KeyVault

### PR DESCRIPTION
**Pull request by tjprescott**
Depends on unmerged PR #7473.

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

**Comment by rakshith91** [comment]
Thanks a lot @tjprescott 
Can you change the base branch to https://github.com/Azure/azure-sdk-for-python/tree/storage-samples-readme

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+def make_resource_name(prefix):
+    return '{}{}'.format(prefix, str(uuid.uuid4()).replace('-', ''))
+
+class KeyWrapper:
+    """ Class that fulfills the interface used by the storage SDK's
+        automatic client-side encyrption and decryption routines. """
+
+    def __init__(self, kek, credential):
+        self.kek = kek
+        self.kid = kek._properties._id
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: If `kek` is a `KeyVaultKey`, you want `kek.id`.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        self.algorithm = KeyWrapAlgorithm.aes_256
+
+    def wrap_key(self, key):
+        crypto_client = CryptographyClient(self.kek, credential)
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Where does `credential` come from? You could create the client in the constructor.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        wrapped = crypto_client.wrap_key(key=key, algorithm=self.algorithm)
+        return wrapped.encrypted_key
+
+    def unwrap_key(self, key, algorithm):
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Constructor and `get_key_wrap_algorithm` imply a `KeyWrapper` is bound to a single algorithm, but this takes an algorithm argument?

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+client_secret = get_env_var(CLIENT_SECRET)  # aka password in AzureCLI
+
+# construct a token credential for use by Storage and KeyVault clients.
+credential = ClientSecretCredential(tenant_id, client_id, client_secret)
+secret_client = SecretClient(keyvault_url, credential=credential)
+
+# the secret is url-safe base64 encoded bytes, content type 'application/octet-stream'
+secret = secret_client.get_secret('symmetric-key')
+key_bytes = base64.urlsafe_b64decode(secret.value)
+kvk = KeyVaultKey(key_id=secret.id, key_ops=['unwrapKey', 'wrapKey'], k=key_bytes, kty='oct')
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Our enum for key types, `azure.keyvault.keys.KeyType`, has an `oct` value.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+kek = KeyWrapper(kvk, credential)
+
+storage_client = BlobServiceClient(storage_url, credential=credential)
+container_name = make_resource_name('container')
+blob_name = make_resource_name('blob')
+
+container_client = storage_client.get_container_client(container_name)
+# if kek is used, error: "wrap_key" is missing or invalid
+# if crypto_client is used, error: "get_kid" is missing or invalid
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Vestigial comments?

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+container_client.key_encryption_key = kek
+container_client.create_container()
+container_client.upload_blob(blob_name, 'This is my blob.')
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: There are many like it, but this one is mine.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+USAGE: python client_side_encryption_keyvault.py
+"""
+
+import base64
+import os
+import sys
+import uuid
+
+from azure.common import AzureException
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: unused import

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+"""
+FILE: client_side_encryption_keyvault.py
+
+DESCRIPTION:
+    This sample contains code demonstrating how to conifgure the storage blob
+    service for client side encryption, storing and retrieving the key encryption key
+    (kek) from within Azure KeyVault. This sample requires a service principal be set
+    configured with access to KeyVault, and that the vault contains a 256-bit base64-
+    encoded secret named "symmetric-key". Additionally, a number of environment
+    variabled must be set. Since these often contain sensitive information, they
```
Comment (chlowell) [inline sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: variable**s**, would be helpful to list them here.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py (hunk 3 above)
Comment (tjprescott) [reply sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Yes, this is the interface defined by storage. If omitted you get an error: "unwrap_key takes three positional arguments, but two were given."

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py (hunk 5 above)
Comment (tjprescott) [reply sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Ah, yes, I was trying to use the KeyVault primitives directly and neither worked for these reasons. I had to write the KeyWrapper class to conform to the Storage SDK's desired interface.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py (hunk 2 above)
Comment (tjprescott) [reply sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Done.

File: sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py (hunk 3 above)
Comment (chlowell) [reply sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Seems bizarre but if we must implement this interface, I'd ignore the argument and pass `self.algorithm` to `unwrap_key`. If the `KeyWrapper` performs all un/wrapping, and doesn't accept an arbitrary wrap algorithm, accepting an arbitrary unwrap algorithm is just a bug vector.

Comment (bryevdv) [reply sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py]: Is this the spot? If so the comment there claims an exception is expected if the algorithm is not supported
https://github.com/Azure/azure-sdk-for-python/blob/db402b152526a41408c2e64e965a398bd893312e/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/encryption.py#L249-L251

**Comment by tjprescott** [comment]
Closing in favor of PR #8186.

**Pull request closed without merging**